bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieRequestProcessor.java

```diff
@@ -229,6 +230,16 @@ protected void onAddRequestFinish() {
         }
     }
 
+    protected void onAddRequestFinishWithoutUnTrack() {
+        if (addsSemaphore != null) {
+            addsSemaphore.release();
+        }
+    }
+
+    protected void onAddRequestUnTrack() {
+        requestStats.untrackAddRequest();
+    }
```

Comment:
You need to track N add requests at onAddRequestStart, and untrack it when onAddRequestFinish.

Reply:
We batched the add requests, but each request callback one by one. So I tracked N add requests at onAddRequestStart, but I should not call the onAddRequestFinish after all the entries callback reached. We need to release the addsSemaphore when one callback is reached and un-track the add requests after all the entries callback reached.